### PR TITLE
[AArch64] Create set.fpmr intrinsic and assembly lowering

### PR DESCRIPTION
**Pull request by Lukacma**
This patch introduces new llvm.set.fpmr intrinsics for setting value in FPMR register and adds its lowering to series of read-compare-write instructions. This intrinsic will be generated during lowering of FP8 C intrinsics into LLVM-IR introduced in later patch.

***This is an experimental implementation of handling fp8 intriniscs and is likely to change in the future.***

### CODE REVIEW

**Review by davemgreen — COMMENTED** [review]
Can you explain how these are intended to be used? I am maybe a bit out of touch but I was under the impression that the FP8 intrinsics had fpmr operands to make data-flow analysis possible. Does a stand-alone intrinsic not go against that?

File: llvm/test/CodeGen/AArch64/arm64-fpenv.ll [hunk 1]
```diff
@@ -37,6 +37,20 @@ define void @set_fpsr(i64 %sr) {
...
 }
 
+define dso_local void @set_fpmr(i64 %sr) {
+; CHECK-LABEL: set_fpmr:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mrs x8, FPMR
+; CHECK-NEXT:    cmp x8, x0
+; CHECK-NEXT:    b.eq .LBB4_2
+; CHECK-NEXT:  // %bb.1:
+; CHECK-NEXT:    msr FPMR, x0
```
[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply llvm/test/CodeGen/AArch64/arm64-fpenv.ll:47]: Hi David,

I agree with you that this does not make sense by itself. It looks like an overkill
But this intrinsic will be used together with FP8 intrinsics.
Each C intrinsic that has an mfloat8_t type will  have a value  for  fpm, that will mapped to  this llvm.set.fmpr intrinsic followed to the instruction intrinsic.

 svfloat16x2_t svcvt1_f16[_mf8]_x2_fpm(svmfloat8_t zn, fpm_t fpm) 

So each fp8 llvm intrinsic will be preceded  by an llvm.set.fpmr.

The write to FPMR is more expensive than comparing and writing only if it is needed. Moreover we believe this value will not change much. But we still dont have real benchmarks using it to know for sure.

I hope this helps.

**Comment by CarolineConcatto** [comment]
> Can you explain how these are intended to be used? I am maybe a bit out of touch but I was under the impression that the FP8 intrinsics had fpmr operands to make data-flow analysis possible. Does a stand-along intrinsic not go against that?

Do you mean that the compiler would have a pass to improve the writes into FMPR?Is that what you mean?
ATM this pass/data-flow analysis does not exist and the compiler would generate a write to FMPR to every FP8 intrinsic.
For instance the example I shown :
svfloat16x2_t svcvt1_f16[_mf8]_x2_fpm(svmfloat8_t zn, fpm_t fpm)
Will lower to: 
llvm.set.fpmr(i64 fpm)
{<vscale x 8 x half>,<vscale x 8 x half>}llvm.aarch64.scvt2.nxv8i16(<vscale x 8 x i8> %zn)
One of the reasons to have an llvm-ir  intrinsic for FPMR instead of  lowering to machine instruction late in the pipeline is because the compiler could hoist llvm.set.fprm outside a loop vectorizer, if it is a constant/does not change. AFAIU to hoist a machine instructions outside a loop is more complicated than llvm-i intrinsics.

File: llvm/test/CodeGen/AArch64/arm64-fpenv.ll [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
-; RUN: llc -mtriple=aarch64 < %s | FileCheck %s
+; RUN: llc -mtriple=aarch64 -mattr=+fpmr -verify-machineinstrs < %s | FileCheck %s
```
Comment (SpencerAbson) [inline llvm/test/CodeGen/AArch64/arm64-fpenv.ll:2]: Does this flag have any meaning upstream?

**Comment by davemgreen** [comment]
> > Can you explain how these are intended to be used? I am maybe a bit out of touch but I was under the impression that the FP8 intrinsics had fpmr operands to make data-flow analysis possible. Does a stand-along intrinsic not go against that?
> 
> Do you mean that the compiler would have a pass to improve the writes into FMPR?Is that what you mean? ATM this pass/data-flow analysis does not exist and we did not tried LICM.
> 
> For this implementation the compiler would generate a write to FMPR to every FP8 intrinsic. For instance the example I shown : svfloat16x2_t svcvt1_f16[_mf8]_x2_fpm(svmfloat8_t zn, fpm_t fpm) Will lower to: llvm.set.fpmr(i64 fpm) {<vscale x 8 x half>,<vscale x 8 x half>}llvm.aarch64.scvt2.nxv8i16(<vscale x 8 x i8> %zn) One of the reasons to have an llvm-ir intrinsic for FPMR instead of lowering to machine instruction late in the pipeline is because the compiler could hoist llvm.set.fprm outside a loop vectorizer, if it is a constant/does not change. AFAIU to hoist a machine instructions outside a loop is more complicated than llvm-i intrinsics.

Hi. Yeah sorry there were two issues I had.
 - The first is that the lowers to read + branch + set, as opposed to just set. This feels like a premature optimization to me, considering this is very new and we don't know whether that will be a sensible optimization or not when there is real hardware. My guess would be that it often hurts more than it help, so unless you have a very strong reason to add it you might consider simplifying it to a simple store, like other operations. We can always add it back in later if needed.
 - The other was a bit of a drive-by comment about the dataflow between the llvm.set.fpmr and the use. I agree that having a separate llvm.set.fpmr is a really sensible idea - that llvm can optimize the placement of and cse. We probably don't want to leave that until very late and rely on machine ir optimizations. But throwing away which fpmr values is acting on the current intrinsic throws away all the info about the types it is using and how the instruction operates. It essentially turns any optimization that you might want to do in the mid-end into a (function-)global optimization, as opposed to keeping it local. It would need to search back through the program to find what the current fpmr might be. It would feel like keeping that intact would be useful, with something like
`%fpmr = i64 llvm.set.fpmr(i64 fpm)`
`{<vscale x 8 x half>,<vscale x 8 x half>}llvm.aarch64.scvt2.nxv8i16(<vscale x 8 x i8> %zn, i64 %fpmr)`
Whether they are kept using Inaccessible memory I'm not sure, it might be useful to keep. It's ugly and prevents a lot of optimizations, but if it prevents overlapping ranges of fpmr then maybe that is a good thing. We would usually use it for things we don't have any other alternative as we have no data-fow information, but there is only a single register in the hardware. It is should be easy to remove later if we find we can, adding a new %fpmr parameter is more difficult.

File: llvm/test/CodeGen/AArch64/arm64-fpenv.ll (hunk 2 above)
Comment (Lukacma) [reply llvm/test/CodeGen/AArch64/arm64-fpenv.ll]: No. Fixed

File: llvm/test/CodeGen/AArch64/arm64-fpenv.ll (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/test/CodeGen/AArch64/arm64-fpenv.ll:43]: Changed to simple msr instrucion.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 3]
```diff
@@ -2145,6 +2145,12 @@ def MSR_FPSR : Pseudo<(outs), (ins GPR64:$val),
                PseudoInstExpansion<(MSR 0xda21, GPR64:$val)>,
                Sched<[WriteSys]>;
 
+let Uses = [FPMR], Defs = [FPMR, NZCV] in
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:2148]: With the simpler usage can this be just `let Defs = [FPMR]`?

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 4]
```diff
@@ -2145,6 +2145,12 @@ def MSR_FPSR : Pseudo<(outs), (ins GPR64:$val),
...
+def SET_FPMR : Pseudo<(outs), (ins GPR64:$val),
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:2149]: I think it best to match the existing naming scheme, which would make this `def MSR_FPMR`.

File: llvm/lib/CodeGen/LivePhysRegs.cpp [hunk 5]
```diff
@@ -262,7 +262,7 @@ void llvm::addLiveIns(MachineBasicBlock &MBB, const LivePhysRegs &LiveRegs) {
   const MachineRegisterInfo &MRI = MF.getRegInfo();
   const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
   for (MCPhysReg Reg : LiveRegs) {
-    if (MRI.isReserved(Reg))
+    if (TRI.getReservedRegs(MF).test(Reg))
```
Comment (paulwalker-arm) [inline llvm/lib/CodeGen/LivePhysRegs.cpp:265]: Can you explain why this change is required?

File: llvm/include/llvm/IR/IntrinsicsAArch64.td [hunk 6]
```diff
@@ -778,6 +778,9 @@ def int_aarch64_get_fpcr : FPENV_Get_Intrinsic;
 def int_aarch64_set_fpcr : FPENV_Set_Intrinsic;
 def int_aarch64_get_fpsr : FPENV_Get_Intrinsic;
 def int_aarch64_set_fpsr : FPENV_Set_Intrinsic;
+def int_aarch64_set_fpmr : DefaultAttrsIntrinsic<[], [llvm_i64_ty], [IntrInaccessibleMemOnly]>{
```
Comment (paulwalker-arm) [inline llvm/include/llvm/IR/IntrinsicsAArch64.td:781]: Perhaps just add an `FPMR_Set_Intrinsic` class to sit along side the existing FPENV ones.

**Comment by Lukacma** [comment]
Hello David,

Thank you for your feedback! I have now fixed the first issue you had with the patch. As for the 2nd one, for now I don't think creating data dependencies is necessary. Functionally, dependencies between defs and uses of FPMR are going to be handled using memory dependency through inaccessible memory. I do not see a "nice" way to handle overlaping ranges  without using memory dependency, but if you know of a way I would be keen on hearing about it and we might change to data dependency only later. I agree that it will help with optimization to not have to search through the program for where the fpmr value might be coming from, but that is something we will think about once we start optimizing these intrinsics.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 3 above)
Comment (Lukacma) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Yes. forgot to remove this when switching to simple write. Fixed

File: llvm/lib/CodeGen/LivePhysRegs.cpp (hunk 5 above)
Comment (Lukacma) [reply llvm/lib/CodeGen/LivePhysRegs.cpp]: No longer required with simple MSR write. Removed.

**Review by paulwalker-arm — APPROVED** [review]

**Comment by davemgreen** [comment]
> Hello David,
> 
> Thank you for your feedback! I have now fixed the first issue you had with the patch. As for the 2nd one, for now I don't think creating data dependencies is necessary. Functionally, dependencies between defs and uses of FPMR are going to be handled using memory dependency through inaccessible memory. I do not see a "nice" way to handle overlaping ranges without using memory dependency, but if you know of a way I would be keen on hearing about it and we might change to data dependency only later. I agree that it will help with optimization to not have to search through the program for where the fpmr value might be coming from, but that is something we will think about once we start optimizing these intrinsics.

As you ask here I will try and reply - we had a conversation about this internally and it seemed the response was that we plan to look into this more and hopefully change it later.  (My worry with that is that is will hard to do the more time is spent with them working this way).  As for handling overlapping ranges it would just be a case of sinking the intrinsic to the use, it sounds like this will need a pass to do the placement in either case. But I don't actually mean to say that it needs to remove the memory side effect entirely.  It could keep the side effect to limit how much the intrinsics move and have the benefits of knowing the fpmr and no overlapping ranges at the same time. You can even remove one of them later if it is decided they are not useful (going the other way is obviously more difficult).

Thanks for simplifying the codegen though - hopefully we won't end up needing it in the future, and if we do we can always resurrect it.
